pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,550 @@
+	// local $HOSTNAME
+	localHostname string
+	// hostname from K8S, could be overridden
+	localK8sHostname string
+	// Photon Project ID
+	projID string
+	cloudprovider.Zone
+}
+
+type PCConfig struct {
+	Global struct {
+		CloudTarget       string `gcfg:"target"`
+		IgnoreCertificate bool   `gcfg:"ignoreCertificate"`
+		Tenant            string `gcfg:"tenant"`
+		Project           string `gcfg:"project"`
+		OverrideIP        bool   `gcfg:"overrideIP"`
+	}
+}
+
+// Disks is interface for manipulation with PhotonController Persistent Disks.
+type Disks interface {
+	// AttachDisk attaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	AttachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DetachDisk detaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	DetachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DiskIsAttached checks if a disk is attached to the given node.
+	DiskIsAttached(pdID string, nodeName k8stypes.NodeName) (bool, error)
+
+	// CreateDisk creates a new PD with given properties.
+	CreateDisk(volumeOptions *VolumeOptions) (pdID string, err error)
+
+	// DeleteDisk deletes PD.
+	DeleteDisk(pdID string) error
+}
+
+// VolumeOptions specifies capacity, tags, name and flavorID for a volume.
+type VolumeOptions struct {
+	CapacityGB int
+	Tags       map[string]string
+	Name       string
+	Flavor     string
+}
+
+func logError(msg string, err error) error {
+	s := "Photon Cloud Provider: " + msg + ". Error [" + err.Error() + "]"
+	glog.Errorf(s)
+	return fmt.Errorf(s)
+}
+
+func readConfig(config io.Reader) (PCConfig, error) {
+	if config == nil {
+		err := fmt.Errorf("cloud provider config file is missing. Please restart kubelet with --cloud-provider=photon --cloud-config=[path_to_config_file]")
+		return PCConfig{}, err
+	}
+
+	var cfg PCConfig
+	err := gcfg.ReadInto(&cfg, config)
+	return cfg, err
+}
+
+func init() {
+	cloudprovider.RegisterCloudProvider(ProviderName, func(config io.Reader) (cloudprovider.Interface, error) {
+		cfg, err := readConfig(config)
+		if err != nil {
+			return nil, logError("failed to read in cloud provider config file", err)
+		}
+		return newPCCloud(cfg)
+	})
+}
+
+// Retrieve the Photon VM ID from the Photon Controller endpoint based on the node name
+func getVMIDbyNodename(project string, nodeName string) (string, error) {
+	vmList, err := Esxclient.Projects.GetVMs(project, nil)
+	if err != nil {
+		return "", fmt.Errorf("Failed to GetVMs from project %s with nodeName %s, error: [%v]", project, nodeName, err)
+	}
+
+	for _, vm := range vmList.Items {
+		if vm.Name == nodeName {
+			if vm.State == "STARTED" {
+				return vm.ID, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("No matching started VM is found with name %s", nodeName)
+}
+
+// Retrieve the Photon VM ID from the Photon Controller endpoint based on the IP address
+func getVMIDbyIP(project string, IPAddress string) (string, error) {
```

Comment:
This is an extremely expensive function. Each call to get the VM IP addresses is expensive (you make an API caller to Photon Controller, which makes an API call to hostd on the appropriate host, and that call isn't cheap).  If there are a lot of VMs, this will be extremely expensive. (Many seconds) How often is this called? 

We should figure out if you can cache the results or some other alternative approach. Can you remind me why you need to do this, and when exactly it's done?